File: jp-meridian9/client/vega_context.lua
-- ============================================================
-- ヴェガ選択肢メニュー（自前 NUI）
-- ox_lib の lib.registerContext は幅・文字サイズを Lua から変えられないため、
-- 任務受注の選択肢のみ本リソースの NUI で表示する。
-- ============================================================

MRD9 = MRD9 or {}

local State = {
    open = false,
    handlers = {},
    onBack = nil,
}

function MRD9.VegaContextHide()
    State.open = false
    State.handlers = {}
    State.onBack = nil
    SetNuiFocus(false, false)
    SendNUIMessage({ type = 'vegaContextClose' })
end

---@param def { title: string, showBack?: boolean, onBack?: function|nil, options: { title: string, description?: string, icon?: string, disabled?: boolean, onSelect?: function }[] }
function MRD9.VegaContextShow(def)
    MRD9.VegaContextHide()

    local serial = {}
    local handlers = {}
    for i, o in ipairs(def.options or {}) do
        handlers[i] = o.onSelect
        serial[i] = {
            title = o.title or '',
            description = o.description,
            icon = o.icon,
            disabled = o.disabled and true or false,
        }
    end

    local scale = 2.0
    if Config and Config.NPC and type(Config.NPC.contextMenuScale) == 'number' and Config.NPC.contextMenuScale > 0 then
        scale = Config.NPC.contextMenuScale
    end

    State.handlers = handlers
    State.onBack = def.onBack
    State.open = true

    SendNUIMessage({
        type = 'vegaContextOpen',
        title = def.title or '',
        showBack = def.showBack and true or false,
        options = serial,
        scale = scale,
    })
    SetNuiFocus(true, true)
end

RegisterNUICallback('mrd9_vega_ctx_select', function(data, cb)
    local idx = tonumber(data and data.index)
    local fn = idx and State.handlers[idx]
    MRD9.VegaContextHide()
    if fn then
        fn()
    end
    cb({ ok = true })
end)

RegisterNUICallback('mrd9_vega_ctx_back', function(_, cb)
    local backFn = State.onBack
    MRD9.VegaContextHide()
    if backFn then
        backFn()
    end
    cb({ ok = true })
end)

RegisterNUICallback('mrd9_vega_ctx_close', function(_, cb)
    MRD9.VegaContextHide()
    cb({ ok = true })
end)

AddEventHandler('onResourceStop', function(res)
    if res ~= GetCurrentResourceName() then
        return
    end
    MRD9.VegaContextHide()
end)

File: jp-meridian9/html/vega_context.js
(function () {
  var root = document.getElementById('mrd9-vega-ctx');
  if (!root) {
    return;
  }

  var backdrop = root.querySelector('.mrd9-vega-ctx__backdrop');
  var panel = root.querySelector('.mrd9-vega-ctx__panel');
  var btnBack = root.querySelector('.mrd9-vega-ctx__back');
  var btnClose = root.querySelector('.mrd9-vega-ctx__close');
  var titleEl = root.querySelector('.mrd9-vega-ctx__title');
  var listEl = root.querySelector('.mrd9-vega-ctx__list');

  var ICON_GLYPH = {
    play: '\u25B6',
    briefcase: '\u25A0',
    gavel: '\u2696',
    'door-open': '\u2302',
    'arrow-right': '\u2192',
    times: '\u2715',
    'file-signature': '\u2261',
    clock: '\u231A',
    check: '\u2713',
    user: '\u25CF',
    users: '\u25CF\u25CF',
    'info-circle': '\u24D8',
  };

  function resName() {
    if (typeof GetParentResourceName === 'function') {
      return GetParentResourceName();
    }
    return 'jp-meridian9';
  }

  function post(action, body) {
    return fetch('https://' + resName() + '/' + action, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json; charset=UTF-8' },
      body: JSON.stringify(body || {}),
    }).catch(function () {});
  }

  function iconGlyph(name) {
    if (!name || typeof name !== 'string') {
      return '\u00A0';
    }
    return ICON_GLYPH[name] || '\u00A0';
  }

  function hide() {
    root.classList.add('mrd9-hidden');
    root.setAttribute('aria-hidden', 'true');
    listEl.innerHTML = '';
  }

  function show(data) {
    var scale = typeof data.scale === 'number' && data.scale > 0 ? data.scale : 2;
    panel.style.setProperty('--mrd9-vega-ctx-scale', String(scale));

    titleEl.textContent = data.title || '';
    if (data.showBack) {
      btnBack.classList.remove('mrd9-hidden');
    } else {
      btnBack.classList.add('mrd9-hidden');
    }

    listEl.innerHTML = '';
    var opts = data.options || [];
    for (var i = 0; i < opts.length; i++) {
      (function (idx, o) {
        var li = document.createElement('button');
        li.type = 'button';
        li.className = 'mrd9-vega-ctx__opt';
        if (o.disabled) {
          li.classList.add('mrd9-vega-ctx__opt--disabled');
          li.disabled = true;
        }

        var ic = document.createElement('span');
        ic.className = 'mrd9-vega-ctx__opt-icon';
        ic.textContent = iconGlyph(o.icon);
        ic.setAttribute('aria-hidden', 'true');

        var body = document.createElement('div');
        body.className = 'mrd9-vega-ctx__opt-body';
        var t = document.createElement('div');
        t.className = 'mrd9-vega-ctx__opt-title';
        t.textContent = o.title || '';
        body.appendChild(t);
        if (o.description) {
          var d = document.createElement('div');
          d.className = 'mrd9-vega-ctx__opt-desc';
          d.textContent = o.description;
          body.appendChild(d);
        }

        li.appendChild(ic);
        li.appendChild(body);
        li.addEventListener('click', function () {
          if (li.disabled) {
            return;
          }
          post('mrd9_vega_ctx_select', { index: idx });
        });
        listEl.appendChild(li);
      })(i + 1, opts[i]);
    }

    root.classList.remove('mrd9-hidden');
    root.setAttribute('aria-hidden', 'false');
  }

  window.addEventListener('message', function (e) {
    var msg = e.data || {};
    if (msg.type === 'vegaContextOpen') {
      show(msg);
    } else if (msg.type === 'vegaContextClose') {
      hide();
    }
  });

  if (backdrop) {
    backdrop.addEventListener('click', function () {
      post('mrd9_vega_ctx_close', {});
    });
  }
  if (btnClose) {
    btnClose.addEventListener('click', function () {
      post('mrd9_vega_ctx_close', {});
    });
  }
  if (btnBack) {
    btnBack.addEventListener('click', function () {
      post('mrd9_vega_ctx_back', {});
    });
  }

  document.addEventListener('keydown', function (e) {
    if (e.key !== 'Escape' || root.classList.contains('mrd9-hidden')) {
      return;
    }
    e.preventDefault();
    post('mrd9_vega_ctx_close', {});
  });
})();
